options/ansi/generic/time-stubs.cpp

```diff
@@ -578,10 +579,445 @@ char *ctime_r(const time_t *clock, char *buf) {
+                    tm->tm_wday = product - 1;
+                    break;
+                }
+                case 'V':
+                    break;
+                case 'z':
+                    break;
+                case 'Z':
+                    break;
+                case 's': //end of GNU extensions
+                    break;
+                case 'E': { //locale-dependent date & time representation
+                    state->format_index++;
+                    switch(format[state->format_index]){
+                        case 'c':
+                            break;
+                        case 'C':
+                            break;
+                        case 'x':
+                            break;
+                        case 'X':
+                            break;
+                        case 'y':
+                            break;
+                        case 'Y':
+                            break;
+                        default:
+                            return NULL;
+                    }
+                }
+                case 'O': { //locale-dependent numeric symbols
+                    state->format_index++;
+                    switch(format[state->format_index]){
+                        case 'd':
+                        case 'e':
+                            break;
+                        case 'H':
+                            break;
+                        case 'I':
+                            break;
+                        case 'm':
+                            break;
+                        case 'M':
+                            break;
+                        case 'S':
+                            break;
+                        case 'U':
+                            break;
+                        case 'w':
+                            break;
+                        case 'W':
+                            break;
+                        case 'y':
+                            break;
+                        default:
+                            return NULL;
+                    }
+                }
+                default:
+                    return NULL;
+            }
+        }
+        else if(isspace(format[state->format_index])){
+            while(isspace(input[state->input_index++]));
+            state->input_index--;
+        }
+        else {
+            if(format[state->format_index] != input[state->input_index++])
+                return NULL;
+        }
+        state->format_index++;
+    }
+
+    return (char*)input + state->input_index;
+}
+
+} //anonymous namespace
+
+/*char *strptime(const char *__restrict, const char *__restrict,
 		struct tm *__restrict) {
 	mlibc::infoLogger() << "mlibc: strptime is a stub!" << frg::endlog;
 	return nullptr;
+}*/
+
+char *strptime(const char *__restrict s, const char *__restrict format, struct tm *__restrict tm){
+    struct strptime_internal_state state = {};
+
+    char *result = strptime_internal(s, format, tm, &state);
```

Comment:
Is there a use for having this as a seperate function? It seems to just complicate the code more.

Reply:
Certain directives are equal to a known combination of other directives (`%D`, `%r`, `%R`, `%T`) and some directives are equal to other directives depending on locale (`%x`,  `%X`). Implementing it this way, as a function that simply calls itself to handle such "composite" cases greatly simplifies the control-flow (we wouldn't need any conditional fall-through, for example) and allows it to handle arbitrary directives which are only known at runtime more easily. To me, this was the most logical and simple way to implement this.

Reply:
Okay, I'm not sure I follow. The function takes some inputs, and produces some outputs and saves them to this ``state`` struct. The state is then examined in the main function afterwards. This ``state`` could also just be some local variables afaict, and have the main "body" of the function just modify these local variables.

Reply:
Yes, the issue is that certain inputs expand to a combination of directives. To me the cleanest way to handle this is a recursive function, especially since we do not know all expansions that can take place beforehand.

The state is passed through to keep track of the entire state of the parser, which is also necessary for deciding what to calculate at the end (we wouldn't need to calculate the day of the week if it's already given, for example.) Since the recursive calls can modify the state of the parser, I'm not sure how you'd have them as local variables instead.

Reply:
Ah I see, I understand now. I think still I'd prefer there to be no extra function. The duplicated code is probably not in such large quantities (afaict) that it warrants recursion like this. In this case I'd prefer to focus on readability rather than conciseness.

Reply:
Well, it's not necessarily just code duplication - certain directives "expand" to a combination of locale-dependent directives. I'm not sure how this would be implemented if not with recursion, and besides, I don't think recursion is atypical for parsers.